File: domain/src/main/java/org/cryptomator/domain/usecases/cloud/ByteArrayDataSource.java
package org.cryptomator.domain.usecases.cloud;

import android.content.Context;

import org.cryptomator.util.Optional;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;

public class ByteArrayDataSource implements DataSource {

	public static DataSource from(byte[] bytes) {
		return new ByteArrayDataSource(bytes);
	}

	private final byte[] bytes;

	private ByteArrayDataSource(byte[] bytes) {
		this.bytes = bytes;
	}

	@Override
	public Optional<Long> size(Context context) {
		long size = bytes.length;
		return Optional.of(size);
	}

	@Override
	public InputStream open(Context context) throws IOException {
		return new ByteArrayInputStream(bytes);
	}

	@Override
	public DataSource decorate(DataSource delegate) {
		return delegate;
	}

	@Override
	public void close() throws IOException {
		// do nothing because ByteArrayInputStream need no close
	}
}

File: domain/src/main/java/org/cryptomator/domain/usecases/cloud/ByteArrayDataSource.kt
package org.cryptomator.domain.usecases.cloud

import android.content.Context
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.util.Date

class ByteArrayDataSource private constructor(private val bytes: ByteArray) : DataSource {

	override fun size(context: Context): Long {
		return bytes.size.toLong()
	}

	override fun modifiedDate(context: Context): Date? {
		return null
	}

	@Throws(IOException::class)
	override fun open(context: Context): InputStream {
		return ByteArrayInputStream(bytes)
	}

	override fun decorate(delegate: DataSource): DataSource {
		return delegate
	}

	@Throws(IOException::class)
	override fun close() {
		// do nothing because ByteArrayInputStream need no close
	}

	companion object {

		@JvmStatic
		fun from(bytes: ByteArray): DataSource {
			return ByteArrayDataSource(bytes)
		}
	}
}
